### PR TITLE
Make arch-specific releases use separate images instead of tags on the same image

### PR DESCRIPTION
**Pull request by towca**
This seems to be the current convention in k8s.

### CODE REVIEW

**Comment by towca** [comment]
@hakman, you introduced arch-specific builds to the Makefile. Could you please take a look?
/assign @hakman

**Comment by hakman** [comment]
@towca In general this is true, but not sure if there is support support for pushing manifests using cross repo references. If you really want the `-amd64` and `-arm64` images pushed to the corresponding repos, may be safer to retag with the corresponding suffixed images.

If you look at what is actually pushed to the prod repo from staging, you will see that there are no tags for the `-amd64` and `-arm64` images, just the main one, so cleanup is done by the promoter.
https://console.cloud.google.com/gcr/images/k8s-artifacts-prod/EU/autoscaling/cluster-autoscaler

**Comment by towca** [comment]
The process of promoting images from staging to prod is completely manual. We haven't promoted any arch-specific `cluster-autoscaler` images yet. I want to promote an arm64 image right now, hence the changes. The promotion shouldn't be an issue, the same thing has already been done for Vertical Pod Autoscaler: https://github.com/kubernetes/k8s.io/pull/1318.

File: cluster-autoscaler/Makefile [hunk 1]
```diff
@@ -53,24 +56,24 @@ make-image: make-image-arch-$(GOARCH)
...
-	./push_image.sh ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG}-$*
+	./push_image.sh ${IMAGE}-$*:${TAG}
 
 push-manifest:
-	DOCKER_CLI_EXPERIMENTAL=enabled docker manifest create ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG} \
-		$(addprefix $(REGISTRY)/cluster-autoscaler$(PROVIDER):$(TAG)-, $(ALL_ARCH))
-	DOCKER_CLI_EXPERIMENTAL=enabled docker manifest push --purge ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG}
+	docker manifest create --amend ${IMAGE}:${TAG} $(shell echo $(ALL_ARCH) | sed -e "s~[^ ]*~$(IMAGE)\-&:$(TAG)~g")
+	@for arch in $(ALL_ARCH); do docker manifest annotate --arch $${arch} ${IMAGE}:${TAG} ${IMAGE}-$${arch}:${TAG}; done
+	docker manifest push --purge ${IMAGE}:${TAG}
```
Comment (hakman) [inline cluster-autoscaler/Makefile:76]: There is no need to annotate the image with the arch.
```suggestion
	docker manifest create ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG} \
		$(addprefix $(REGISTRY)/cluster-autoscaler$(PROVIDER)-, $(addsuffix :$(TAG), $(ALL_ARCH)))
	docker manifest push --purge ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG}
```

File: cluster-autoscaler/Makefile [hunk 2]
```diff
@@ -53,24 +56,24 @@ make-image: make-image-arch-$(GOARCH)
 make-image-arch-%:
 ifdef BASEIMAGE
 	docker build --pull --build-arg BASEIMAGE=${BASEIMAGE} \
-		-t ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG}-$* \
+		-t ${IMAGE}-$*:${TAG} \
 		-f Dockerfile.$* .
 else
 	docker build --pull \
-		-t ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG}-$* \
+		-t ${IMAGE}-$*:${TAG} \
```
Comment (hakman) [inline cluster-autoscaler/Makefile:63]:
```suggestion
		-t ${REGISTRY}/cluster-autoscaler${PROVIDER}-$*:${TAG} \
```

File: cluster-autoscaler/Makefile [hunk 3]
```diff
@@ -53,24 +56,24 @@ make-image: make-image-arch-$(GOARCH)
 make-image-arch-%:
 ifdef BASEIMAGE
 	docker build --pull --build-arg BASEIMAGE=${BASEIMAGE} \
-		-t ${REGISTRY}/cluster-autoscaler${PROVIDER}:${TAG}-$* \
+		-t ${IMAGE}-$*:${TAG} \
```
Comment (hakman) [inline cluster-autoscaler/Makefile:59]:
```suggestion
		-t ${REGISTRY}/cluster-autoscaler${PROVIDER}-$*:${TAG} \
```

File: cluster-autoscaler/Makefile [hunk 4]
```diff
@@ -28,6 +28,9 @@ ifdef DOCKER_RM
 else
   RM_FLAG=
 endif
+IMAGE=$(REGISTRY)/cluster-autoscaler$(PROVIDER)
+
```
Comment (hakman) [inline cluster-autoscaler/Makefile:32]:
```suggestion
```

**Review by hakman — CHANGES_REQUESTED** [review]
Simplified things a bit. My concern was that the push to staging won't work.
Also, if you want to push the ARM64 image in https://github.com/kubernetes/k8s.io/pull/1485, you should also push the AMD64 one.

File: cluster-autoscaler/Makefile (hunk 1 above)
Comment (towca) [reply cluster-autoscaler/Makefile]: Done.

**Comment by towca** [comment]
I've already tested pushing to staging, so that shouldn't be a concern (gcr.io/k8s-staging-autoscaling/cluster-autoscaler-arm64:v1.20.0).

I'm building and promoting the arm64 image for a previous release specifically because somebody asked for it during the SIG meeting. All next releases will have all 3 images released. But sure, I can include the amd64 image as well.

File: cluster-autoscaler/Makefile (hunk 2 above)
Comment (hakman) [reply cluster-autoscaler/Makefile:63]: Seemed easier to read for me like without using the variable and doesn't seem to be that much more room for mistakes in this case. In any case, it's just a preference.

**Comment by hakman** [comment]
Thanks. I tested this also and worked fine for me.
/lgtm

File: cluster-autoscaler/Makefile (hunk 2 above)
Comment (towca) [reply cluster-autoscaler/Makefile:63]: Ack

**Comment by towca** [comment]
Thanks for the review!
/assign @MaciekPytel

**Comment by hakman** [comment]
Thanks for the releases @towca :)

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Pull request merged**